projects/wallet-template/src/background/createBackgroundRpc.ts

```diff
@@ -37,23 +41,50 @@ const keyset = {
 }
 
 const createKeyring = () => {
-  let savedPassword = "123456"
-  let currentPassword: string | undefined
+  const getKeystore = () => storage.get("password")
+  const setKeystore = (keystore: KeystoreV4) =>
+    storage.set("password", keystore)
+  const removeKeystore = () => storage.remove("password")
+  let isLocked = true
+
   return {
-    unlock(password: string) {
-      if (password !== savedPassword) throw new Error("invalid password")
-      currentPassword = password
+    async unlock(password: string) {
+      const keystore = await getKeystore()
+      assert(keystore, "keyring must be setup")
+      if (!keystoreV4.verifyPassword(keystore, password))
+        throw new Error("invalid password")
+      isLocked = false
+    },
+    async lock() {
+      assert(await getKeystore(), "keyring must be setup")
+      isLocked = true
+    },
+    async isLocked() {
+      assert(getKeystore(), "keyring must be setup")
+      return isLocked
     },
-    lock() {
-      currentPassword = undefined
+    async changePassword(currentPassword: string, newPassword: string) {
+      const keystore = await getKeystore()
+      assert(keystore, "keyring must be setup")
+      if (!keystoreV4.verifyPassword(keystore, currentPassword))
+        throw new Error("invalid password")
+      await setKeystore(
+        keystoreV4.create(
+          newPassword,
+          keystoreV4.decrypt(keystore, currentPassword),
+        ),
+      )
+
+      // TODO: re-encrypt accounts with new password
     },
-    isLocked() {
-      return !currentPassword
+    async setup(password: string) {
+      assert(!(await getKeystore()), "keyring is already setup")
+      await setKeystore(keystoreV4.create(password, randomBytes(32)))
```

Comment:
@ryanleecode `randomBytes(32)` is the encrypted secret.
With the password it can be decrypted and used as a key to encrypt seed phrases or private keys.

WDYT?

Reply:
better to use the key incase the cipher changes in the future

Reply:
for the above alternatives, when there is a password change
- keeping the key does not need account secret re-encryption because the key hasn't changed
- using cipher, anything that was encrypted with the cipher needs to be re-encrypted